client/client.go

```diff
@@ -218,12 +221,39 @@ func (cli *Client) ClientVersion() string {
 // UpdateClientVersion updates the version string associated with this
 // instance of the Client. This operation doesn't acquire a mutex.
 func (cli *Client) UpdateClientVersion(v string) {
+	// If no version specified, negotiate with the server
+	if v == "" {
+		cli.DowngradeClientVersion()
+	}
 	if !cli.manualOverride {
 		cli.version = v
 	}
 
 }
 
+// DowngradeClientVersion downgrades the version string associated with this
+// instance of the Client to match the latest version the server supports
+func (cli *Client) DowngradeClientVersion() {
+	ping, _ := cli.Ping(context.Background())
+	cli.DowngradeClientVersionPing(ping)
+}
+
+// DowngradeClientVersionPing downgrades the version string associated with this
+// instance of the Client to match the latest version the server supports
+func (cli *Client) DowngradeClientVersionPing(p types.Ping) {
```

Comment:
Same comment about the function name, how about `UpdateAPIVersionFromPing()` ?

The check for `if !cli.manualOverride {...}` needs to move here. It should be the first line in the function I think:

```go
if cli.manualOverride {
    return
}
```